source/shared/cpp/ObjectModel/Column.cpp

```diff
@@ -132,10 +94,20 @@ std::shared_ptr<Column> Column::Deserialize(ParseContext& context, const Json::V
     column->SetVerticalContentAlignment(ParseUtil::GetEnumValue<VerticalContentAlignment>(
         value, AdaptiveCardSchemaKey::VerticalContentAlignment, VerticalContentAlignment::Top, VerticalContentAlignmentFromString));
 
+    // configures for container style
+    column->ConfigForContainerStyle(context);
+
+    // we walk parse tree dfs in-order, so we need to save current style,
+    // before we walk back up to a parent.
+    context.SaveContextForCollectionTypeElement(column); 
+
     // Parse Items
     auto cardElements = ParseUtil::GetElementCollection(context, value, AdaptiveCardSchemaKey::Items, false);
     column->m_items = std::move(cardElements);
 
+    // since we are walking dfs, we have to restore the style before we back up
+    context.RestoreContextForCollectionTypeElement(column);
```

Comment:
I'm a little uneasy with us having to manually maintain this context. The other stack we're using during parsing is managed through parser wrappers (see e.g. `BaseCardElementParserWrapper` and `ElementParserRegistration::GetParser()`) which make it difficult to get wrong. Also, it feels a little unusual to have two separate stacks to maintain ancestor information at parse time. I'm thinking maybe we should:
* Change `ParseContext::m_idStack` to be `std::vector<std::pair<std::shared_ptr<BaseElement>, bool>> m_elementStack;`
* Add a new method for `BaseElement`: `virtual bool CanHaveChildren() const { return false; }`
* In `CollectionTypeElement`, override `BaseElement::CanHaveChildren()` to return true
* Change bleed logic to walk `m_elementStack` to find the information it needs

This sets us up for the next time we need to refer to ancestors to make parsetime decisions :)

What do you think?

Reply:
thanks paul for guiding me to the right direction. i agree with cons of this manual context saving, so i thinking about adding parser class for CollectionTypeElement, and implement deserialize method that deserializes all of the member variables of CollectionTypeItem, such that it would work similar to parserWrapper.
 but i'm hesitant about using the same stack because the use of separate stacks makes the feature  implementation simple. 


---
In reply to: [259107103](https://github.com/Microsoft/AdaptiveCards/pull/2406#discussion_r259107103) [](ancestors = 259107103)